aten/src/ATen/native/mps/operations/ScanKernel.mm

```diff
@@ -140,14 +141,165 @@ static void scan_mps_impl(const Tensor& self,
   });
 }
 
+// Utility function to get 2D grid dimensions for dispatch
+static std::pair<uint32_t, uint32_t> get_2d_grid_dims(const IntArrayRef& shape, const int64_t dim) {
+  size_t grid_x = 1;
+  size_t grid_y = 1;
+
+  for (int i = 0; i < dim; ++i) {
+    if (grid_x * shape[i] < UINT32_MAX) {
+      grid_x *= shape[i];
+    } else {
+      grid_y *= shape[i];
+    }
+  }
+
+  TORCH_CHECK(grid_y <= UINT32_MAX && grid_x <= UINT32_MAX, "Unable to safely factor shape for grid dimensions.");
+
+  if (grid_y > grid_x) {
+    std::swap(grid_x, grid_y);
+  }
+
+  return std::make_pair(static_cast<uint32_t>(grid_x), static_cast<uint32_t>(grid_y));
+}
+
+// Specialized implementation for cummin/cummax that returns both values and indices
+static void scan_with_indices_mps_impl(const Tensor& self,
+                                       const Tensor& values_output,
+                                       const Tensor& indices_output,
+                                       int64_t dim,
+                                       const std::string& op_name) {
+  if (values_output.numel() == 0) {
+    return;
+  }
+
+  const int64_t ndim = self.dim();
+  const int64_t wrapped_dim = maybe_wrap_dim(dim, ndim);
+  const int64_t axis_size = self.size(wrapped_dim);
+
+  // Preprocess input tensor - ensure it's contiguous for Metal shaders
+  Tensor input_tensor = self.contiguous();
+
+  // Preprocess output tensors - ensure they're contiguous for Metal shaders
+  Tensor values_tensor = values_output;
+  Tensor indices_tensor = indices_output;
+  bool values_needs_copy = !values_output.is_contiguous();
+  bool indices_needs_copy = !indices_output.is_contiguous();
+  Tensor temp_values, temp_indices;
+
+  if (values_needs_copy) {
+    temp_values = at::empty_like(values_output, values_output.options()).contiguous();
+    values_tensor = temp_values;
+  }
+
+  if (indices_needs_copy) {
+    temp_indices = at::empty_like(indices_output, indices_output.options()).contiguous();
+    indices_tensor = temp_indices;
+  }
```

Comment:
This avoids a lots of unnecessary copies (`contiguous()` is a no-op if tensor is already contiguous)

```suggestion
  auto values_tensor = values_output.contiguous();
  auto indices_tensor = indices_output.contiguous();
  const bool values_needs_copy = !values_output.is_contiguous();
  const bool indices_needs_copy = !indices_output.is_contiguous();
```

Reply:
applied!